### ISSUE TITLE
StockItemCriteriaMapper has wrong function declaration

### ISSUE THREAD

**Issue by sydekumf** [magento/magento2#10979]
When you set a products filter with an array of product ids on a stock item criteria and want to get a list of stock items via the stock item repository, you only get one item instead the number of product ids given. This is due to a wrong function declaration.

### Preconditions
1. Create some piece of code you can execute (i.e. a plugin, see code example)

```php
<?php

namespace Vendor\Module\Plugin\MyPlugin;

use Magento\CatalogInventory\Api\StockItemCriteriaInterfaceFactory;
use Magento\CatalogInventory\Api\StockItemCriteriaInterface;
use Magento\CatalogInventory\Api\StockItemRepositoryInterface;

class MyPlugin
{
    /**
     * @var StockItemRepositoryInterface
     */
    protected $stockItemRepository;

    /**
     * @var StockItemCriteriaInterface
     */
    protected $stockItemCriteria;

    public function __construct(
        StockItemRepositoryInterface $stockItemRepository,
        StockItemCriteriaInterfaceFactory $stockItemCriteriaFactory
    ){
        $this->stockItemRepository = $stockItemRepository;
        $this->stockItemCriteria = $stockItemCriteriaFactory->create();
    }

    public function someMethod($productIds)
    {
        $this->stockItemCriteria->setProductsFilter($productIds);
        $stockItems = $this->stockItemRepository->getList($this->stockItemCriteria)->getItems();

        foreach ($stockItems as $stockItem) {
            // do something
        }
    }
}
```

### Steps to reproduce
1. Try to execute the code above or try to debug the code behind `getList()` 
2. You get to `Magento\Framework\DB\AbstractMapper::map()` where you see `call_user_func_array([$this, $mapperMethod], $value)` which will call the function `Magento\CatalogInventory\Model\ResourceModel\Stock\Item\StockItemCriteriaMapper::mapProductsFilter($products)`
3. There is the error: The function expects ONE argument and not many (array)

### Expected result
1. The function declaration should be like this:
``` php
public function mapProductsFilter()
{
    $products = $args = func_get_args();
    // ...
}
```

### Actual result
1. The function declaration is wrong

**Comment by magento-engcom-team** [comment]
@sydekumf, thank you for your report.
We were not able to reproduce this issue by following the steps you provided. Tested on 2.1.9.

1. In Magento\Framework\DB\AbstractMapper::map() 
call_user_func_array([$this, $mapperMethod], $value) is called, where :
- $this = it is class "Magento\CatalogInventory\Model\ResourceModel\Stock\Item\StockItemCriteriaMapper";
- $mappeMethod = "mapProductsFilter";
- $value = array of productIds;
2. Method  mapProductsFilter($products) expects productId or array of product ids.
Next code is located in mapProductsFilter($products): 
`$productIds = [];
        if (!is_array($products)) {
            $products = [$products];
        }`

So there is not any error. Please fix our steps to reproduce if we are wrong.

**Comment by sydekumf** [comment]
This issue is independent from Magento, as you handle the PHP function `call_user_func_array()` wrong. Here is an example to reproduce, I stick to the function declarations from Magento:

```php
<?php

class AbstractMapper
{
    public function map($value)
    {
        $mapperMethod = 'mapProductsFilter';
        call_user_func_array([$this, $mapperMethod], $value);
    }

    public function mapProductsFilter($products)
    {
        echo var_export($products);
    }
}

$mapper = new AbstractMapper();
$mapper->map([1, 2, 3]);
```

The solution would be if you call it like this:
`$mapper->map([[1, 2, 3]]);`
But that is not according to the API declaration of the method:

```php
/**
     * Add product filter to collection
     *
     * @param int|int[] $products
     * @return bool
     */
    public function setProductsFilter($products);
```

You see `int[]`? Then it should be `int[][]`. So that is not a solution, as it does not conform to the API.

Another solution would be to change the `mapProductsFilter` to the following:

```php
public function mapProductsFilter()
{
    $products = $args = func_get_args();
    // ...
}
```

You decide which solution you want to go. I just give you information about the problem and how to reproduce it. Again: This problem is specific to the usage of the PHP function as it is used WRONG in Magento :-)

**Comment by dverkade** [comment]
I will look into this issue and try to fix it. #SQUASHTOBERFEST

**Comment by vrann** [comment]
This is duplicate of #7678

**Comment by okorshenko** [comment]
https://github.com/magento/magento2/issues/7678 fixed in 2.2-develop
